### PR TITLE
Docker + Fig setup

### PR DESCRIPTION
**Pull request by lightsofapollo**
Note- I am not sure when I have time to finish this up for real... I deployed an image (lightsofapollo/treeherder) for testing and it is working really well for my TC related work... This is somewhat janky and aimed more at third party developers though it could be cleaned up.

### CODE REVIEW

**Comment by wlach** [comment]
What's the advantage of this over Vagrant?

**Comment by parkouss** [comment]
Fuck you !

2015-01-12 16:16 GMT+01:00 William Lachance <notifications@github.com>:

**Comment by parkouss** [comment]
I'm sorry for the last comment guys. It went out unintentionally and was certainly not addressed to any of you. I removed it.
Again, I apologize for the inconvenience.

**Comment by lightsofapollo** [comment]
tldr; this is super useful for testing treeherder within another project it is both faster and easier then using multi-vagrant-box setup. 

I am not sure how valuable this is to core treeherder development as it requires a docker setup that does volume mounts of user folder (which requires special configuration for OSX though works out of the box on linux)

(@wlach) To address the "why"  @maurodoglio had mentioned he was dabbling in a fig configuration I figured it would make sense to at least post my work here to see if there is overlap... I think this could be hugely useful for anyone who needs to write tests against treeherder without wanting to hit dev/staging/prod (this is how I use it and it is awesome).

**Comment by wlach** [comment]
Would merging treeherder-ui into treeherder-service solve the volume mount issue?

**Comment by lightsofapollo** [comment]
merging could solve other problems but the volume mount issue here is more that docker does not run on OSX so if you want to volume mount your working directory into the docker container you must be doing some nfs/sharing trickery via vagrant (see https://github.com/lightsofapollo/docker-env).

Assuming you had the above setup correctly then you could volume mount something very similar to how vagrant does it (../treeherder-ui) and it should work just fine.

**Comment by lightsofapollo** [comment]
Related- vagrant has great support for running docker I _think_ right now autophone and tc are the only clients of treeherder but if there where more (in which case docker as a first class thing is more valuable) it is an easy thing to take the above work and make it run _inside_ the vagrant image (effectively removing the need for a chunk of the puppet scripts). 

Originally those puppet scripts where used for both vagrant and staging I am not sure if that still is the case?

**Comment by maurodoglio** [comment]
@lightsofapollo we used those puppet scripts to setup the environment on dev and the previous version of staging. staging is a different beast now.

File: treeherder/etl/management/commands/export_project_credentials.py [hunk 1]
```diff
@@ -53,5 +53,5 @@ def write_credentials(file_path):
 
     immutable_credentials = TreeherderModelBase.get_oauth_credentials()
     keys_fh = open(file_path, 'w')
-    keys_fh.write(json.dumps(immutable_credentials))
+    keys_fh.write(json.dumps(immutable_credentials, indent=4))
```
Comment (lightsofapollo) [inline treeherder/etl/management/commands/export_project_credentials.py]: You will thank me later

File: docker/etc/profile.d/treeherder.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+export TREEHERDER_DJANGO_SECRET_KEY='5up3r53cr3t'
+# Allow any host to be hit when running in development mode
+export TREEHERDER_ALLOWED_HOSTS='*'
+
+# link : mysql
+export TREEHERDER_DATABASE_NAME=$MYSQL_ENV_MYSQL_DATABASE
+export TREEHERDER_DATABASE_USER='root'
+export TREEHERDER_DATABASE_PASSWORD=$MYSQL_ENV_MYSQL_ROOT_PASSWORD
+export TREEHERDER_DATABASE_HOST='mysql'
+export TREEHERDER_DATABASE_PORT='3306'
```
Comment (edmorley) [inline docker/etc/profile.d/treeherder.sh]: Port is unused as of bug 1125530.

File: docker/etc/profile.d/treeherder.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Mapping is taken from docker link assumptions
```
Comment (edmorley) [inline docker/etc/profile.d/treeherder.sh]: Would be good to consolidate some of this, https://github.com/mozilla/treeherder-service/blob/master/puppet/manifests/vagrant.pp#L39 and https://github.com/mozilla/treeherder-service/blob/master/puppet/manifests/production.pp#L40 at some point in the future.

File: docker/generate_test_credentials.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# Load and process the json...
+content = json.loads(open(CREDENTIALS).read())
+
+for project in content.keys():
+    content[project] = {
+        "consumer_key":  project,
+        "consumer_secret":  project,
+    }
+
+cred_file = open(CREDENTIALS, 'w')
```
Comment (wlach) [inline docker/generate_test_credentials.py]: This could be rewritten as:

with open(CREDENTIALS, 'w') as cred_file:
    cred_file.write(...)

Comment (lightsofapollo) [inline docker/generate_test_credentials.py]: /me happy to let gc do the work but will rewrite

Comment (lightsofapollo) [inline docker/generate_test_credentials.py]: ah I did not let gc do the work anyway so nicer
